contracts/zaps/ZapSteth.sol

```diff
@@ -0,0 +1,309 @@
+            _zeroIfPrincipal,
+            1
+        );
+    }
+
+    /// @notice There should never be any tokens in this contract.
+    /// This function can rescue any possible leftovers.
+    /// @param token The token to rescue.
+    /// @param amount The amount to rescue.
+    function rescueTokens(address token, uint256 amount) external onlyOwner {
+        if (token == address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)) {
+            amount = Math.min(address(this).balance, amount);
+            payable(msg.sender).transfer(amount);
+        } else {
+            IERC20 want = IERC20(token);
+            amount = Math.min(want.balanceOf(address(this)), amount);
+            want.transfer(msg.sender, amount);
+        }
+    }
+
+    /// @notice Helper function that takes principal or yield tokens and redeems them for ETH or stETH.
+    /// @param _amount Amount of PT or YT to withdraw.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _slippageAllowance The slippage allowed for the internal curve stable-swap call.
+    /// @param _zeroIfPrincipal Zero if the assets to redeem are principal tokens.
+    /// @param _zeroIfEth Zero if the assets to redeem for is ETH. otherwise it will be stETH.
+    function _zapOut(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _slippageAllowance,
+        int128 _zeroIfPrincipal,
+        int128 _zeroIfEth
+    ) internal {
+        ITranche tranche = _deriveTranche(address(_position), _expiration);
+
+        if (_zeroIfPrincipal == 0) {
+            tranche.transferFrom(msg.sender, address(this), _amount);
+            tranche.approve(address(tranche), _amount);
+            tranche.withdrawPrincipal(_amount, address(this));
+        } else {
+            IERC20 yt = IERC20(tranche.interestToken());
+            yt.transferFrom(msg.sender, address(this), _amount);
+            yt.approve(address(tranche), _amount);
+            tranche.withdrawInterest(_amount, address(this));
+        }
+
+        uint256 balance = want.balanceOf(address(this));
+        require(balance > 0, "no balance");
+
+        _noReentry = true;
+        // balance - burn amount
+        // _zeroIfEth - coin withdraw index
+        // 0 - minimum amount of coin to receive
+        StableSwapSTETH.remove_liquidity_one_coin(balance, _zeroIfEth, 0);
+        _noReentry = false;
+
+        uint256 endBalance;
+        if (_zeroIfEth == 0) {
+            endBalance = address(this).balance;
+            payable(msg.sender).transfer(endBalance);
+        } else {
+            endBalance = stETH.balanceOf(address(this));
+            stETH.transfer(msg.sender, endBalance);
+        }
+        require(
+            (endBalance * (_slippageAllowance + 10000)) / 10000 >= balance,
+            "TOO MUCH SLIPPAGE"
+        );
+
+        uint256 leftover = tranche.balanceOf(address(this));
```

Comment:
Our leftover checks only apply to pt not yt? Shouldn't this also be switched on the `_zeroIfPrincipal` factor